ident2.8
.TH ident2 8 GNU
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
ident2 \- An advanced ident daemon
.SH SYNOPSIS
.B ident2
.I "[options]"
.SH "DESCRIPTION"
ident2 is an advanced ident daemon, which can run as an inetd service or
on it's own. It has various features such as user defined and random
replies, extensive logging, and full administrative control.
.PP
.SH OPTIONS
The program follows the usual GNU syntax command-line unless it is
unsupported on your system. GNU options begin with two dashes(`-') and have
much more descriptive names.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-u, \-\-usage
Shows example command lines
.TP
.B \-v, \-\-version
Show version information
.TP
.B \-l, \-\-license
Show licensing information
.TP
.B \-a, \-\-force\-inetd
Force inetd mode, if autodetection isn't working.
.TP
.B \-d, \-\-force\-daemon
Force standalone daemon, if autodetection isn't working.
.TP
.B \-i, \-\-use\-user\-ident
Allow user defined ident replies. Just do a
.B echo 'ident ar' > ~/.ident
to change your ident reply.
.TP
.B \-y, \-\-user\-reply\-file
File ident2 should look for in user's homedir for replies. Default is ~/.ident
.TP
.B \-n, \-\-allow\-noident\-file
Disable replies for user if user has a ~/.noident file. This behavior is disabled by default.
.TP
.B \-o, \-\-client\-timeout
Clients timeout after this many seconds
.TP
.B \-s, \-\-dont\-change\-uid
Ident2 will normally try to change it's uid to something less privileged if
it finds it's uid to be 0. This option disables this behavior altogether.
.TP
.B \-r, \-\-always\-random
Always send a random reply, regardless of what kind of request the client
made. This is useful for abusing the ident protocol as well as assisting 
outgoing IP masqueraded connections.
.TP
.B \-m, \-\-daemon\-maxclients
Daemon-only: Accept no more than this many connections at a time.
.TP
.B \-p, \-\-daemon\-port
Daemon-only: Bind this port, instead of 'auth' aka 113
.TP


.SH "SEE ALSO"
.IR "inetd(8), inetd.conf(5)" 
.SH AUTHOR
This manual page was written by Alexander Reelsen <ar@rhwd.net>,
for the Debian GNU/Linux system (but may be used by others).


CHANGES
--------- [07/22/2005
+ incorporated lsof patch which adds support for
  SunOS, NetBSD, OpenBSD, OSF/1, HP-UX
  (thanks inittab@netizenweb.com (Greg Schenzel))

--------- [05/18/2004
+ buffer overflow in _getl fixed :(

--------- [11/02/2003
+ fixed using data after it has been freed in sys/m_fbsd.c:m_get_uid()
  thanks to Ilya Novoselov <nullguid@t72.ru>

--------- [08/18/2000
+ forgot to specifiy option 'n' in getopt_long
  thanks to Johan Ekenberg <johan@ekenberg.se> for pointing this out

--------- [04/11/2000
+ correctly handles errno == EINTR when accept() returns -1
+ date formats in CHANGES file now Y2K compliant!

--------- [11/2/99
+ port to FreeBSD
+ PID registration, under FreeBSD, is also /var/run

--------- [9/18/99
+ changed all sprintf's to snprintf's
+ removed all potential security hazards (strcpy, memcpy, sprint, etc)
+ moved m_register_pid AFTER the initial fork()-to-become-daemon,
  otherwise wrong pid is registered.
+ d'oh. move m_reduce_rights AFTER m_register_pid, otherwise can't
  write pid to /var/run
- NOTE, snprintf is not portable. must be removed in future.

--------- [9/3/99]
+ performing security sweep. reevaluating choices
+ merged getopt_long support
+ added getopt_long detection to ./configure
+ wrote two help outputs (short and short+long versions)
+ implemented the --license parameter
+ implemented the --usage paramater

--------- [8/30/99]
+ moved srand(time(NULL)) out of INIT_GLOBALS and into the random reply generator
  function. rand is seeded every time, which is flawed, but still harmless.
  this was moved because the parent never actually calls rand(), which sucks
  because the children will keep generating the same reply.
+ implemented new machine call, m_register_pid(), for use in daemon mode,
  on systems that wish to keep track of running system services.
+ PID registration, under Linux, in /var/run
+ switched to a ./configure based setup

--------- [8/25/99]
+ changed m_linux.c to only setuid(). Under Linux, setuid() sets the effective
  userid as well as superuser. If this would be the case, seteuid() should fail.

--------- [8/9/99]
+ added random replies
+ fixed standard unspecial replies (INIT_GLOBALS wasn't initting!)
- ripped out config file parsers
- ripped out all semblances of IP masquerading
+ rather than trying to drop userid to nobody and failing if it
  doesn't work, it only tries to change uid if uid is 0, elsewise,
  keep current uid.
+ reworded README
  
--------- [6/11/99]
- removed throttled daemon. I felt this included some very unnatural
  code and it was a pain to debug. it was a research project at most.
+ rewrote config file parser
+ rewrote command-line processor
+ included GNU information on top of all source files
+ code cleaning spree. Optimized, minimalized, etc.
+ fixed buffer handling in some areas which may have been dangerous
+ reorganized some general code into common.c
+ stablized threaded daemon
+ rewrote README
+ reorganized throttled daemon
+ moved accept() out into the child process so we don't have
  to do any nasty stuff like return file descriptors to the
  parent through exit().
+ adopted capitalize global variables convention
+ decided to start DATING these changes

---------- 0.999c
+ no idea why I thought the user of 'nobody' couldn't carry on identity
  functions. The dropping to nobody under Linux has been reenacted.
+ fixed a diverge from RFC1413. Thanks to Piotr Kucharski <chopin@sgh.waw.pl>
  for pointing out that END OF LINE is clearly defined by the RFC as CRLF
  and not LF. Apologies to anyone affected by this.
+ fixed an 'off by one' error in the throttled implementation causing a
  garbage char to appear at the end of an ident query.
+ command-line option to display compile-time options.
+ shamelessly copied a pidentd convention. If you have the file .noident
  in your home directory, it will not send a response to anyone requesting
  an ident that resolves to your username.
+ dropping userid to 'nobody' has been changed to instead drop the access
  level to a user defined variable (which SHOULD be a username).
+ moved quite a few variables out into seperate #defines. 
+ dropping userid to 'nobody' as a security measure is back, again.
+ IP masquerading works! Much love to everyone who helped the
  debugging/fixing process. Especially Hans.
  -Actually, it only sort of works. Linux 2.0 and 2.1 have
  differing levels of success. Kernel 2.2 is supposedly going to
  settle on a format once and for all.
+ rand() wasn't being seeded. 
+ for some reason I wasn't making use of #define USER_REPLIES.
+ dropping the userid to 'nobody' makes servicing ident requests
  impossible. I'm dumb. The function 'go_nobody' and all references
  to it were removed.
+ major MAJOR rewrite. This introduced some more natural (ie, not kludgy)
  code as well as more bugs! It should also be easier to extend and
  expand. Hopefully noone will get hurt this time. :)



ident2.c
/* 
 * Ident-2 - an Identity server for UNIX
 * Copyright (C) 1998-2001 Michael Bacarella
 * Copyright (C) 2003 Netgraft Corporation
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Please view the file README for program information.
 *
 */

#include "ident2.h"

#ifdef HAS_GETOPT_LONG
#include <getopt.h>	/* getopt() is declared in unistd.h */
#endif

	/*
	 *	determine if we're running as a child of inetd
	 *	or if we were started via a command line (or from
	 *	a process not inetd)
	 */
static int
inetd_child (void)
{
	struct sockaddr_in sin;
	int sinsize = sizeof (struct sockaddr_in);
	
		/* if child of inetd, 0 would be a socket */
	if (getsockname (0, (struct sockaddr *)&sin, (int *)&sinsize) == -1) {
		if (errno == ENOTSOCK)
			return 0;	/* must be a filedescriptor */
		else {
			syslog (LOG_ERR, "getsockname: %s\n", strerror (errno));
			exit (-1);
		}
	}
	return 1;
}

#define print_header()	printf ("Ident2 version %s, Copyright (C) 1999-2004 Michael Bacarella\n", ID_VERSION)

static void
print_usage (void)
{
	print_header();
	puts ("Example command lines:");
puts ("		ident2 -r		sends random replies, always");
puts ("		ident2 -i -y .reply	allows user replies, which are read\n"
      "					from the file ~/.reply");
puts ("		ident2 -m 5 -o 10	ultra secure mode. no more than\n"
      "					5 connections at once, each connection\n"
      "					is killed after 10 seconds regardless.");
puts ("		ident2 -s		stop ident2 from trying to change it's\n"
      "					userid if it discovers that it's running\n"
      "					as a root process.");
}

static void
print_license (void)
{
	print_header();
	puts(
"ident2 is free software, and you are welcome to redistribute it\n"
"under certain conditions; ident2 comes with ABSOLUTELY NO WARRANTY;\n"
"for details, go to http://www.gnu.org/\n");
}

static void
print_help (void)
{
	print_header();
#ifdef HAS_GETOPT_LONG
	puts (  
"usage: ident2 [options]\n"
"options:\n"
"\n"
"these parameters apply to all incarnations of ident2\n"
"\n"
"	-h	--help			this command line information\n"
"	-u	--usage			shows example command lines\n"
"	-v	--version		show version information\n"
"	-l	--license		show licensing information\n"
"	-a	--force-inetd		force inetd mode\n"
"	-d	--force-daemon		force standalone daemon\n"
"	-i	--use-user-ident	allow user defined ident replies\n"
"	-y	--user-reply-file	file in user's homedir for replies\n"
"	-n	--allow-noident-file	don't reply if user has a ~/NOIDENT file\n"
"	-o	--client-timeout	clients timeout after this many secs\n"
"	-s	--dont-change-uid	don't try to change uid (to nobody)\n"
"	-r	--always-random		always send a random reply\n"
"\n"
"these parameters only apply to the daemon ident2\n"
"\n"
"	-m	--daemon-maxclients	accept no more than this many clients\n"
"	-p	--daemon-port		bind this port, instead of 'auth'\n"
);
#else
	puts(
"usage: ident2 [options]\n"
"options:\n"
"\n"
"these parameters apply to all incarnations of ident2\n"
"\n"
"	-h		this command line information\n"
"	-u		shows example command lines\n"
"	-v		show version information\n"
"	-l		show licensing information\n"
"	-a		force inetd mode\n"
"	-d		force standalone daemon\n"
"	-i		allow user defined ident replies\n"
"	-y		file in user's homedir for replies\n"
"	-n		don't reply if user has a ~/NOIDENT file\n"
"	-o		clients timeout after this many secs\n"
"	-s		don't try to change uid (ex: to nobody)\n"
"	-r		always send a random reply, even to bad requests\n"
"\n"
"these parameters only apply to the daemon ident2\n"
"\n"
"	-m		accept no more than this many clients\n"
"	-p		bind this port, instead of 'auth'\n"
);
#endif
}

        /*
         *      print version info
         */
static void
print_version (void)
{
        printf ("Ident2 version %s, maintainer %s\n\n",
                ID_VERSION, ID_MAINTAINER);
}


static void
command_line (int argc, char *argv[])
{
	int c;
			/* getopt_long support implemented
			 * by Alexander Reelsen <ar@rhwd.net> */
#ifdef HAS_GETOPT_LONG
	while (1) {
/* int this_option_optind = optind ? optind : 1; */	/*unused?*/
		int option_index = 0;
		static struct option long_options[] = {
			{"allow-noident-file", 0, 0, 'n'},
			{"help", 0, 0, 'h'},
			{"usage", 0, 0, 'u'},
			{"version", 0, 0, 'v'},
			{"license", 0, 0, 'l'},
			{"force-inetd", 0, 0, 'a'},
			{"force-daemon", 0, 0, 'd'},
			{"use-user-ident", 0, 0, 'i'},
			{"dont-change-uid", 0, 0, 's'},
			{"always-random", 0, 0, 'r'},
			{"user-reply-file", 1, 0, 'y'},
			{"client-timeout", 1, 0, 'o'},
			{"daemon-maxclients", 1, 0, 'm'},
			{"daemon-port", 1, 0, 'p'},
			{0, 0, 0, 0},
		};

		c = getopt_long (argc, argv, "nhuvrliady:o:ftTsm:p:",
			long_options, &option_index);
		if (c == -1) {
			break;
		}	
#else
        while ((c = getopt (argc, argv, "nhuvrliady:o:ftTsm:p:")) != -1) {
#endif
		switch (c) {
			case 'n':
				Allow_NOIDENT = TRUE;
				break;
			case 'h':
				print_help ();
				exit (0);
			case 'u':
				print_usage ();
				exit (0);	
			case 'v':
				print_version ();
				exit (0);
			case 'l':
				print_license ();
				exit (0);
			case 'a':
				Service_Type = INETD;
				break;
			case 'd':
				Service_Type = DAEMON;
				break;
			case 'i':
				Use_User_Ident = TRUE;
				break;
			case 's':
				Dont_Change_Uid = TRUE;
				break;
			case 'r':
				Reply_Always_Random = TRUE;
				break;
			case 'y':
				User_Ident_File = optarg;
				break;
			case 'o':
				if ((Client_Timeout = atol (optarg)) == 0) {
					fprintf (stderr, "ERROR: Ident2: bad "
						"value for arguement `o'\n");
					exit (1);
				}
				break;
			case 'm':
				if ((Max_Connections = atol (optarg)) == 0) {
					fprintf (stderr, "ERROR: Ident2: bad"
						" value for arguement `m'\n");
					exit (1);
				}
				break;
			case 'p':
				if ((Ident_Port = atoi (optarg)) == 0) { 
					fprintf (stderr, "ERROR: Ident2: bad"
						" value for arguement `p'\n");
					exit (1);
				}
				break;
			case '?':
			default:
#ifdef HAS_GETOPT_LONG
				fprintf (stderr, "ident2: bad command line:\n"
				"\t``ident2 --help | more'' for help\n"
				"\t``ident2 --usage'' for some examples\n\n");
#else
				fprintf (stderr, "ident2: bad command line:\n"
				"\t``ident2 -h | more'' for help\n"
				"\t``ident2 -u'' for some examples\n\n");
#endif
				exit (1);
		}
	}
}


int
main (int argc, char *argv[])
{	
	INIT_GLOBALS();
	command_line (argc, argv);
	 	
			/* try to determine the service type if nothing
			 * has been forced (by commandline) */
	if (Service_Type == NO_TYPE) {
		if (inetd_child ())
			Service_Type = INETD;
		else
			Service_Type = DAEMON;
	}

	if (Service_Type == INETD) {
		openlog ("in.ident2", LOG_PID, LOG_AUTH);
		child_service (STDIN_FILENO, STDOUT_FILENO);
		closelog ();
	}
	else
		daemon_service ();

	return 0;
}

